lib/charms/operator_libs_linux/v1/systemd.py

```diff
@@ -66,149 +71,146 @@
 class SystemdError(Exception):
     """Custom exception for SystemD related errors."""
 
-    pass
 
+def _systemctl(*args: str, check: bool = False) -> Optional[int]:
+    """Control a system service using systemctl.
 
-def _popen_kwargs():
-    return {
-        "stdout": subprocess.PIPE,
-        "stderr": subprocess.STDOUT,
-        "bufsize": 1,
-        "universal_newlines": True,
-        "encoding": "utf-8",
-    }
+    Args:
+        *args: Arguments to pass to systemctl.
+        check:
+            Check the output of the systemctl command. Default: False.
+
+    Raises:
+        SystemdError:
+            Raised if returncode != 0 and check is True.
+            None will be returned instead of a returncode.
+    """
+    cmd = ["systemctl", *args]
+    logger.debug(f"Executing command: {cmd}")
+    try:
+        proc = subprocess.run(
+            [*cmd],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+            bufsize=1,
+            encoding="utf-8",
+            check=check,
+        )
+        logger.debug(
+            f"Command {cmd} exit code: {proc.returncode}. systemctl output:\n{proc.stdout}"
+        )
+        return proc.returncode
+    except subprocess.CalledProcessError as e:
+        raise SystemdError(
+            f"Command {cmd} failed with returncode {e.returncode}. systemctl output:\n{e.stdout}"
+        )
 
 
-def _systemctl(
-    sub_cmd: str, service_name: str = None, now: bool = None, quiet: bool = None
-) -> bool:
-    """Control a system service.
+def service_running(service_name: str) -> bool:
+    """Report whether a system service is running.
 
     Args:
-        sub_cmd: the systemctl subcommand to issue
-        service_name: the name of the service to perform the action on
-        now: passes the --now flag to the shell invocation.
-        quiet: passes the --quiet flag to the shell invocation.
+        service_name: the name of the service to check
     """
-    cmd = ["systemctl", sub_cmd]
-
-    if service_name is not None:
-        cmd.append(service_name)
-    if now is not None:
-        cmd.append("--now")
-    if quiet is not None:
-        cmd.append("--quiet")
-    if sub_cmd != "is-active":
-        logger.debug("Attempting to {} '{}' with command {}.".format(cmd, service_name, cmd))
+    # If returncode is 0, this means that is service is active.
+    if _systemctl("--quiet", "is-active", service_name) == 0:
```

Comment:
Simpler (and arguably clearer?) to just do `return _systemctl("--quiet", "is-active", service_name) == 0`. Similar below.

Reply:
Implemented in commit https://github.com/canonical/operator-libs-linux/pull/106/commits/c951f68b8fe18cbab6fc69858e838283063c1e48